### ISSUE TITLE
Some tests that catch `NullReferenceException` hang on Linux under GCStress=0xC

### ISSUE THREAD

**Issue by pgavlin** [dotnet/coreclr#12224]
The following tests exhibit this behavior:
- tests/src/Interop/MarshalAPI/ReadWrite/ReadWriteByte.csproj
- tests/src/Interop/MarshalAPI/ReadWrite/ReadWriteInt16.csproj
- tests/src/Interop/MarshalAPI/ReadWrite/ReadWriteInt32.csproj
- tests/src/Interop/MarshalAPI/ReadWrite/ReadWriteInt64.csproj
- tests/src/Interop/MarshalAPI/ReadWrite/ReadWriteIntPtr.csproj
- tests/src/JIT/Regression/CLR-x86-JIT/V1-M10/b08944/b08944b.csproj

These tests are being temporarily disabled while the hang is investigated. It appears that the runtime is entering an infinite loop when searching for an appropriate exception handler when an AV in a callee is turned into an NRE.

cc @janvorli

**Comment by pgavlin** [comment]
@janvorli can you take a look at this?

**Comment by janvorli** [comment]
@pgavlin sure!

**Comment by gkhanna79** [comment]
Are we tracking this for 2.0?

**Comment by janvorli** [comment]
I don't think it needs to be tracked for 2.0. It is an issue with GC stress 0xC only and from what I've seen so far, it seems to be caused by some bad interaction between the handling of the SIGSEGV caused by the intentional access violation in the test and the GCStress 0xC access violation handling for the inserted invalid instructions that are used to trigger GC after each jitted instruction.

**Comment by janvorli** [comment]
@gkhanna79 I've found the cuprit and fixed it locally. But I have found that the issue can also happen when e.g. a NullReferenceException happens in catch handler of another NullReferenceException. It is a problem caused by the Linux kernel reusing storage for the signal context. 

So having found that, my answer to your question now changes - I believe we should fix it for 2.0.

What happens is below:
* hardware exception happens
* signal handler is called and it receives a pointer to the signal context (context of the hardware exception)
* we propagate that exception in runtime, find a catch handler, but that catch handler code causes hardware exception again.
* signal handler is called and it receives a pointer to the signal context - the same location as before
* we do first pass of the exception handling. When we hit the frame of our common signal handler function, we don't use the libunwind to go further, but we instead get pointer to the signal context that we have stored to a local  when we were executing the common signal handler and set the unwound context to that context.
* when we are hit the common signal handler frame for the secondary exception, everything works fine
* but later we hit the common signal handler frame for the primary exception and due to the issue I have described at the beginning, the signal context is the context of the secondary exception and so the unwinder jumps "back"
* and it ends up spinning in an infinite loop until the end of time

**Comment by janvorli** [comment]
Minimal repro that hangs forever:
```C#
    class Program
    {
        static void Main(string[] args)
        {
            string s = null;
            try
            {
                int x = s.Length;
            }
            catch (NullReferenceException e)
            {
                int x = s.Length;
            }
        }
    }
```

**Comment by gkhanna79** [comment]
Thanks @janvorli. Is it fair to say this is an issue with 1.0.x and 1.1.x as well? What is the risk profile of the fix?

**Comment by janvorli** [comment]
@gkhanna79 this issue didn't exit in 1.0 and 1.1, it is a regression in 2.0. I've just verified that. It was introduced by my change #7440.

**Comment by janvorli** [comment]
@gkhanna79 as for the risk profile, there is no risk.  It just fixes the case when we hang, but doesn't influence anything else.

**Comment by janvorli** [comment]
Fixed by  #12316

**Comment by gkhanna79** [comment]
@janvorli Given the above, I think you should send a 2.0.0 PR to shiproom to get this ported for PReview3.

**Comment by janvorli** [comment]
@gkhanna79 - I've just done that.

**Comment by janvorli** [comment]
Really fixed by #12344
